Add public API implementation and wire it up to the rewriter.

333fred opened the pull request:
This adds an initial implementation of the API and hooks it up the rewriter information, and addresses a few prototype comments (while also adding a few more). We'll need to have a more fine-grained control over whether to enable the nullable analysis than the global context option, but I'm using that for now as it's a simple solution and works well for our tests. I've hooked up the API to our `VerifyTypes` hook, which is now using entirely public APIs instead of manually binding nodes and running the walker. @AlekseyTs specifically for review, as this is doing a bunch of stuff in the semantic model. @dotnet/roslyn-compiler for review.

cston commented on src/Compilers/CSharp/Portable/BoundTree/BoundNode.cs:
```diff
@@ -170,29 +170,19 @@ public void ResetCompilerGenerated(bool newCompilerGenerated)
+                // This is broken out into a separate property so the debugger can display the
+                // top level nullability without setting the _attributes flag and interferring
+                // with the normal operation of tests.
+                return TopLevelNullabilityDebuggerDisplay;
```
>TopLevelNullabilityDebuggerDisplay [](start = 23, length = 34)

This property is not just for debugger display, it is the core value. Perhaps name it `TopLevelNullabilityCore` instead.

cston commented on src/Compilers/CSharp/Portable/BoundTree/BoundNode.cs:
```diff
@@ -170,29 +170,19 @@ public void ResetCompilerGenerated(bool newCompilerGenerated)
         /// Top level nullability for the node. This should not be used by flow analysis.
         /// </summary>
+        [DebuggerHidden]
         protected NullabilityInfo TopLevelNullability
```
>NullabilityInfo [](start = 18, length = 15)

Consider exposing the internal `NullableAnnotation` and `NullableFlowState` values and have the `SemanticModel` be the one place where internal values are mapped to the public enum values.

333fred replied:
I prefer to do it this way.

---
In reply to: [269156210](https://github.com/dotnet/roslyn/pull/34429#discussion_r269156210) [](ancestors = 269156210)

cston replied:
Ideally it should be possible to implement the analysis without converting from the public enum to the internal enum.

---
In reply to: [269762381](https://github.com/dotnet/roslyn/pull/34429#discussion_r269762381) [](ancestors = 269762381,269156210)

cston commented on src/Compilers/CSharp/Portable/BoundTree/NullabilityRewriter.cs:
```diff
@@ -8,5 +13,70 @@ protected override BoundExpression VisitExpressionWithoutStackGuard(BoundExpress
+
+            while (true)
+            {
+                if (stack.Count == 0)
```
`while (stack.Count > 0)`

cston commented on src/Compilers/CSharp/Portable/BoundTree/NullabilityRewriter.cs:
```diff
@@ -8,5 +13,70 @@ protected override BoundExpression VisitExpressionWithoutStackGuard(BoundExpress
+                currentBinary = updateNode(currentBinary, left, this, nodeUpdater);
+            }
+
+            static BoundBinaryOperatorBase updateNode(BoundBinaryOperatorBase node, BoundExpression leftOpt, NullabilityRewriter rewriter, Func<BoundBinaryOperatorBase, BoundExpression, BoundExpression, TypeSymbol, BoundBinaryOperatorBase> nodeUpdater)
```
>updateNode [](start = 43, length = 10)

Perhaps make this an instance method, to avoid the `rewriter` argument.

333fred replied:
I'd prefer to keep this as a local function, as it's only relevant to this function.

---
In reply to: [269160348](https://github.com/dotnet/roslyn/pull/34429#discussion_r269160348) [](ancestors = 269160348)

cston commented on src/Compilers/CSharp/Portable/BoundTree/UnboundLambda.cs:
```diff
@@ -118,6 +119,7 @@ public TypeWithAnnotations GetInferredReturnType(ConversionsBase conversions, Nu
                 var compilation = Binder.Compilation;
                 NullableWalker.Analyze(compilation,
                                        lambda: this,
+                                       (Conversions)conversions,
```
>(Conversions) [](start = 39, length = 13)

Can the method argument type be changed to `Conversions`?

333fred replied:
No. `NullableWalker` needs a non-null `Conversions` in order to correctly analyze code.

---
In reply to: [269162419](https://github.com/dotnet/roslyn/pull/34429#discussion_r269162419) [](ancestors = 269162419)

cston replied:
As we discussed offline, the question was whether the argument to this method could be declared `Conversions` rather than `ConversionsBase`.

---
In reply to: [269763362](https://github.com/dotnet/roslyn/pull/34429#discussion_r269763362) [](ancestors = 269763362,269162419)

cston commented on src/Compilers/CSharp/Portable/BoundTree/UnboundLambda.cs:
```diff
@@ -191,7 +193,7 @@ public TypeWithAnnotations GetInferredReturnType(ConversionsBase conversions, Nu
                     if (conversions.IncludeNullability)
                     {
-                        bestResultType = NullableWalker.BestTypeForLambdaReturns(returns, compilation, node);
+                        bestResultType = NullableWalker.BestTypeForLambdaReturns(returns, compilation, node, (Conversions)conversions);
```
>(Conversions) [](start = 109, length = 13)

Can the method argument be changed to `Conversions`?

333fred replied:
Added a prototype comment to run on all tests and throw the results away, as I think there might be a case where this will throw.

---
In reply to: [269163031](https://github.com/dotnet/roslyn/pull/34429#discussion_r269163031) [](ancestors = 269163031)

cston commented on src/Compilers/CSharp/Portable/Compilation/CSharpSemanticModel.cs:
```diff
@@ -571,22 +576,22 @@ private static SymbolInfo GetSymbolInfoFromSymbolOrNone(ITypeSymbol type)
         {
             var variable = GetDeclaredSymbol(variableDesignation, cancellationToken);
 
             if (variable != null)
```
>if (variable != null) [](start = 12, length = 21)

No longer necessary.

cston commented on src/Compilers/CSharp/Portable/Compilation/CSharpSemanticModel.cs:
```diff
@@ -1927,16 +1932,17 @@ private static bool IsUserDefinedTrueOrFalse(BoundUnaryOperator @operator)
-                TypeSymbol type = null, convertedType = null;
+                TypeSymbol type = null, convertedType;
+                NullabilityInfo nullability = boundExpr.TopLevelNullability, convertedNullability;
                 Conversion conversion;
```
>Conversion conversion [](start = 16, length = 21)

Consider moving `convertedType`, `convertedNullability`, and `conversion` declarations down to next `if`.

cston commented on src/Compilers/CSharp/Portable/Compilation/CSharpSemanticModel.cs:
```diff
@@ -1985,26 +1994,32 @@ private static bool IsUserDefinedTrueOrFalse(BoundUnaryOperator @operator)
                     {
                         var convertedTuple = (BoundConvertedTupleLiteral)tupleLiteralConversion.Operand;
                         type = convertedTuple.NaturalTypeOpt;
+                        nullability = convertedTuple.TopLevelNullability;
```
Consider extracting a local function:
```
(TypeSymbol, NullabilityInfo) getTypeAndNullability(BoundExpression expr) => ...;
```

cston commented on src/Compilers/CSharp/Portable/Compilation/MemberSemanticModel.cs:
```diff
@@ -1833,10 +1836,40 @@ internal ImmutableArray<BoundNode> GetBoundNodes(CSharpSyntaxNode node)
+                    using (_nodeMapLock.DisposableWrite())
+                    {
+                        _ = GuardedAddBoundTreeAndGetBoundNodeFromMap(node, root);
+                    }
```
Please consider moving this after the `if`, to share the update across both branches.

cston commented on src/Compilers/CSharp/Portable/Compilation/MemberSemanticModel.cs:
```diff
@@ -1833,10 +1836,40 @@ internal ImmutableArray<BoundNode> GetBoundNodes(CSharpSyntaxNode node)
+            }
+            else
+            {
+                results = GuardedAddBoundTreeAndGetBoundNodeFromMap(node, root);
```
>GuardedAddBoundTreeAndGetBoundNodeFromMap [](start = 26, length = 41)

Why is `DisposableWrite` used above but not here?

333fred replied:
Because I have a bug :)

---
In reply to: [269185380](https://github.com/dotnet/roslyn/pull/34429#discussion_r269185380) [](ancestors = 269185380)

cston commented on src/Compilers/CSharp/Portable/SymbolDisplay/SymbolDisplay.cs:
```diff
@@ -67,15 +75,25 @@ public static class SymbolDisplay
         public static string ToMinimalDisplayString(
             ITypeSymbol symbol,
-            CodeAnalysis.NullableFlowState topLevelNullability,
+            CodeAnalysis.NullableFlowState nullableFlowState,
```
>nullableFlowState [](start = 43, length = 17)

In the methods above, this parameter is `topLevelNullability`. Should the parameter names match?

cston commented on src/Compilers/CSharp/Portable/SymbolDisplay/SymbolDisplay.cs:
```diff
@@ -128,25 +157,48 @@ public static class SymbolDisplay
         {
-            format = format ?? SymbolDisplayFormat.MinimallyQualifiedFormat;
-            return ToDisplayParts(symbol, topLevelNullability, semanticModel, position, format, minimal: true);
+            var annotation = (CodeAnalysis.NullableAnnotation?)new TypeWithState((TypeSymbol)symbol, nullableFlowState.ToInternalFlowState()).ToTypeWithAnnotations().NullableAnnotation.ToPublicAnnotation();
```
>TypeSymbol [](start = 82, length = 10)

Is the type guaranteed to be a `TypeSymbol`?

cston commented on src/Compilers/CSharp/Portable/SymbolDisplay/SymbolDisplayVisitor.cs:
```diff
@@ -94,9 +94,19 @@ private static string EscapeIdentifier(string identifier)
+            VisitWithAnntoation(TypeWithAnnotations.Create(symbol, topLevelAnnotation));
+        }
+
+        private void VisitWithAnntoation(TypeWithAnnotations type)
```
>Anntoation [](start = 30, length = 10)

Typo

cston commented on src/Compilers/CSharp/Portable/Symbols/NullableAnnotationExtensions.cs:
```diff
@@ -52,5 +53,28 @@ internal static CodeAnalysis.NullableAnnotation ToPublicAnnotation(this Nullable
+                return default;
+            }
+
+            NullableAnnotation internalAnnotation = (NullableAnnotation)annotation - 1;
```
>NullableAnnotation internalAnnotation = (NullableAnnotation)annotation - 1; [](start = 12, length = 75)

Consider using a `switch` instead, to avoid depending on the actual values of the internal enum.

cston commented on src/Compilers/CSharp/Portable/Symbols/NullableAnnotationExtensions.cs:
```diff
@@ -52,5 +53,28 @@ internal static CodeAnalysis.NullableAnnotation ToPublicAnnotation(this Nullable
+        internal static NullableAnnotation ToInternalAnnotation(this CodeAnalysis.NullableAnnotation annotation)
+        {
+            Debug.Assert((CodeAnalysis.NullableAnnotation)(NullableAnnotation.Oblivious + 1) == CodeAnalysis.NullableAnnotation.Disabled);
+            return annotation == CodeAnalysis.NullableAnnotation.NotApplicable ? NullableAnnotation.Oblivious : (NullableAnnotation)(annotation - 1);
```
>CodeAnalysis.NullableAnnotation.NotApplicable ? NullableAnnotation.Oblivious : (NullableAnnotation)(annotation - 1) [](start = 33, length = 115)

Consider using a `switch` instead, to avoid depending on the actual values of the internal enum.

cston commented on src/Compilers/CSharp/Portable/Symbols/NullableAnnotationExtensions.cs:
```diff
@@ -52,5 +53,28 @@ internal static CodeAnalysis.NullableAnnotation ToPublicAnnotation(this Nullable
+            return new NullabilityInfo(annotation.ToPublicAnnotation(), flowState.ToPublicFlowState());
+        }
+
+        internal static NullableAnnotation ToInternalAnnotation(this CodeAnalysis.NullableAnnotation annotation)
```
>ToInternalAnnotation [](start = 43, length = 20)

Is this method necessary? (Do we need to convert from the public enum to the internal enum?)

333fred replied:
The symbol display API needs it currently. Hopefully after refactoring, it'll convert internal values to public and we won't need this.

---
In reply to: [269208962](https://github.com/dotnet/roslyn/pull/34429#discussion_r269208962) [](ancestors = 269208962)

cston commented on src/Compilers/CSharp/Portable/Symbols/NullableAnnotationExtensions.cs:
```diff
@@ -52,5 +53,28 @@ internal static CodeAnalysis.NullableAnnotation ToPublicAnnotation(this Nullable
             Debug.Assert((CodeAnalysis.NullableAnnotation)(NullableAnnotation.Oblivious + 1) == CodeAnalysis.NullableAnnotation.Disabled);
             return (CodeAnalysis.NullableAnnotation)annotation + 1;
```
>(CodeAnalysis.NullableAnnotation)annotation + 1 [](start = 19, length = 47)

Consider using a `switch` instead.

333fred replied:
I've been resisting doing this, but you and @jcouv have protested enough that I'll cave.

---
In reply to: [269209292](https://github.com/dotnet/roslyn/pull/34429#discussion_r269209292) [](ancestors = 269209292)

cston commented on src/Compilers/CSharp/Test/Semantic/Semantics/NullableReferenceTypesTests.cs:
```diff
@@ -9293,7 +9293,7 @@ public class Base
 }
 ";
-            var compilation = CreateCompilation(new[] { source });
+            var compilation = CreateCompilation(new[] { source }, options: WithNonNullTypesTrue());
```
>options: WithNonNullTypesTrue() [](start = 66, length = 31)

Is this necessary?

333fred replied:
With my changes to `VerifyTypes`, it now uses the public API to get type info instead of running the walker manually. Currently, whether or not the nullable analysis is performed is gated on whether the global nullable context is disabled or not. For the tests that use `VerifyTypes` that didn't have this set, I went through and made changes to the ones that I didn't think did this deliberately, and disabled the ones that I thought were relying on the global context being disabled so that we can re-enable later.

---
In reply to: [269210594](https://github.com/dotnet/roslyn/pull/34429#discussion_r269210594) [](ancestors = 269210594)

cston commented on src/Compilers/CSharp/Test/Semantic/Semantics/NullableReferenceTypesTests.cs:
```diff
@@ -32405,31 +32409,33 @@ void F(bool b, object o1, object? o2, C<object> o3, C<object?> o4)
     static C<object> o5 = null;
 }";
-            var comp = CreateCompilation(new[] { source });
+            var comp = CreateCompilation(new[] { source }, options: WithNonNullTypesTrue());
```
>options: WithNonNullTypesTrue() [](start = 59, length = 31)

We probably should not change the tests. If these changes are needed for the tests to pass, please skip the tests instead, with a PROTOTYPE comment.

cston commented on src/Compilers/Core/CodeAnalysisTest/Symbols/NullablilityInfoTests.cs:
```diff
@@ -0,0 +1,26 @@
+            Assert.NotEqual(new NullabilityInfo(NullableAnnotation.Annotated, NullableFlowState.MaybeNull),
+                            new NullabilityInfo(NullableAnnotation.Annotated, NullableFlowState.NotNull));
+#pragma warning restore IDE0055 // Fix formatting
+        }
```
>} [](start = 8, length = 1)

Consider testing `GetHashCode()` as well.

cston commented:
> using System;

Missing copyright.

---
Refers to: src/Compilers/CSharp/Test/Symbol/Symbols/Source/NullablePublicAPITests.cs:1 in 366d40f. [](commit_id = 366d40fe2c1744ff52d5c488bac3e06e27e038c9, deletion_comment = False)

cston commented on src/Compilers/CSharp/Portable/SymbolDisplay/SymbolDisplay.cs:
```diff
@@ -37,11 +37,19 @@ public static class SymbolDisplay
             return ToDisplayParts(symbol, topLevelNullability, format).ToDisplayString();
         }
 
+        public static string ToDisplayString(
```
>ToDisplayString [](start = 29, length = 15)

Are there tests for the new `SymbolDisplay` methods?

333fred replied:
Not currently. I want to come back and refactor these, as we really need to be able to handle non `TypeSymbol` inputs. I'm just doing the simple thing for now, but forgot to put a prototype. I'll add one

---
In reply to: [269216952](https://github.com/dotnet/roslyn/pull/34429#discussion_r269216952) [](ancestors = 269216952)

cston commented on src/Compilers/CSharp/Portable/SymbolDisplay/SymbolDisplayVisitor.cs:
```diff
@@ -94,9 +94,19 @@ private static string EscapeIdentifier(string identifier)
 
         public void VisitWithNullability(TypeSymbol symbol, NullableFlowState topLevelNullability)
```
>TypeSymbol [](start = 41, length = 10)

Is it still necessary to use `TypeSymbol` in `SymbolDisplay` rather than `ITypeSymbol`, etc.?

333fred replied:
Currently yes, but after a refactor I have planned no.

---
In reply to: [269217827](https://github.com/dotnet/roslyn/pull/34429#discussion_r269217827) [](ancestors = 269217827)

cston commented on src/Compilers/CSharp/Test/Symbol/Symbols/Source/NullablePublicAPITests.cs:
```diff
@@ -0,0 +1,634 @@
+                PublicNullableAnnotation.Annotated);
+        }
+
+        private void VerifyAcrossCompilations<T>(string source,
```
>void [](start = 16, length = 4)

`static`

cston commented on src/Compilers/CSharp/Test/Symbol/Symbols/Source/NullablePublicAPITests.cs:
```diff
@@ -0,0 +1,634 @@
+                for (int i = 0; i < expectedNullabilities.Length; i++)
+                {
+                    Assert.Equal(expectedNullabilities[i], nullabilityFunc(members[i]));
+                }
```
`AssertEx.Equal(members.Select(nullabilityFunc), expectedNullabilities);` will report a diff of the members. And an `itemInspector:` arg can format the result if needed.

cston commented on src/Compilers/Test/Utilities/CSharp/CompilationTestUtils.cs:
```diff
@@ -306,31 +306,19 @@ internal static void VerifyTypes(this CSharpCompilation compilation, SyntaxTree
+                    var typeInfo = model.GetTypeInfo(annotation.Expression);
+                    Assert.NotEqual(CodeAnalysis.NullableAnnotation.NotApplicable, typeInfo.Nullability.Annotation);
+                    Assert.NotEqual(CodeAnalysis.NullableFlowState.NotApplicable, typeInfo.Nullability.FlowState);
+                    return new TypeWithState((TypeSymbol)typeInfo.Type, (NullableFlowState)(typeInfo.Nullability.FlowState - 1)).ToTypeWithAnnotations().ToDisplayString(TypeWithAnnotations.TestDisplayFormat);
```
>new TypeWithState((TypeSymbol)typeInfo.Type, (NullableFlowState)(typeInfo.Nullability.FlowState - 1)).ToTypeWithAnnotations() [](start = 27, length = 125)

Is this conversion needed or could we call the following?
```
typeInfo.Type.ToDisplayString(
    typeInfo.Nullability.FlowState,
    TypeWithAnnotations.TestDisplayFormat);
```

333fred replied:
Unfortunately, part of the test formatting code currently lives in `TypeWithAnnotations.ToDisplayString`. Correcting this is part of my planned refactor of the symbol display apis.

---
In reply to: [269223710](https://github.com/dotnet/roslyn/pull/34429#discussion_r269223710) [](ancestors = 269223710)

AlekseyTs commented on src/Compilers/CSharp/Portable/Compilation/MemberSemanticModel.cs:
```diff
@@ -1435,19 +1435,22 @@ private CSharpSyntaxNode GetBindingRoot(CSharpSyntaxNode node)
-                }
+                // PROTOTYPE(nullable-api): This doesn't work for speculative models: the root is just the statement
+                // or expression being speculated on.
+                return Root;
```
>return Root; [](start = 16, length = 12)

It would be good to add a comment explaining why is this the right thing to return the ```Root``` in this case.

AlekseyTs commented on src/Compilers/CSharp/Portable/Compilation/CSharpSemanticModel.cs:
```diff
@@ -123,6 +123,11 @@ internal static bool CanGetSemanticInfo(CSharpSyntaxNode node, bool allowNamedAr
+        /// <summary>
+        /// True if we should enable nullable analysis on information returned from this semantic model.
+        /// </summary>
+        internal bool EnableNullableAnalysis => Compilation.Options.NullableContextOptions != NullableContextOptions.Disable;
```
>Compilation.Options.NullableContextOptions != NullableContextOptions.Disable [](start = 48, length = 76)

It feels like this is the wrong condition to check. We should simply check if the feature is enabled, i.e. check the language version.

333fred replied:
I'm leaving this as is for now, and we can resolve in the thread we have going on the subject.

---
In reply to: [269786876](https://github.com/dotnet/roslyn/pull/34429#discussion_r269786876) [](ancestors = 269786876)

AlekseyTs commented on src/Compilers/CSharp/Portable/Compilation/CSharpSemanticModel.cs:
```diff
@@ -571,22 +576,22 @@ private static SymbolInfo GetSymbolInfoFromSymbolOrNone(ITypeSymbol type)
         /// figure out its type by looking at the declared symbol of the corresponding variable.
         /// </summary>
-        private ITypeSymbol TypeFromVariable(SingleVariableDesignationSyntax variableDesignation, CancellationToken cancellationToken)
+        private (ITypeSymbol Type, CodeAnalysis.NullableAnnotation annotation) TypeFromVariable(SingleVariableDesignationSyntax variableDesignation, CancellationToken cancellationToken)
```
>(ITypeSymbol Type, CodeAnalysis.NullableAnnotation annotation) [](start = 16, length = 62)

Inconsistent capitalization in member names

AlekseyTs commented on src/Compilers/CSharp/Portable/Compilation/CSharpSemanticModel.cs:
```diff
@@ -1927,16 +1932,17 @@ private static bool IsUserDefinedTrueOrFalse(BoundUnaryOperator @operator)
+                    switch (boundExpr)
                     {
-                        case BoundKind.Local:
+                        case BoundLocal local:
```
>case BoundLocal local: [](start = 24, length = 22)

I am not sure what is the motivation for switching on a type.

333fred replied:
Like the other case, I would need to save the `BoundLocal` off to a local variable to get at the nullability, so I refactored this to have that local introduced for the scope of the relevant case.

---
In reply to: [269792020](https://github.com/dotnet/roslyn/pull/34429#discussion_r269792020) [](ancestors = 269792020)

AlekseyTs commented on src/Compilers/CSharp/Portable/Compilation/CSharpSemanticModel.cs:
```diff
@@ -571,22 +576,22 @@ private static SymbolInfo GetSymbolInfoFromSymbolOrNone(ITypeSymbol type)
-                        return ((ILocalSymbol)variable).Type;
-                    case SymbolKind.Field:
-                        return ((IFieldSymbol)variable).Type;
+                    case ILocalSymbol local:
```
>case ILocalSymbol local: [](start = 20, length = 24)

I am not sure what is the motivation for switching on a type.

333fred replied:
For both of these cases, I would have had to store the cast to a local variable so I can get at both the type and annotation, and all of Andy and Neal's benchmarking here has indicated that this method is just as fast as the kind test, while being shorter and removing an extra `if` that becomes unecessary.

---
In reply to: [269792200](https://github.com/dotnet/roslyn/pull/34429#discussion_r269792200) [](ancestors = 269792200)

AlekseyTs commented on src/Compilers/CSharp/Portable/Compilation/MemberSemanticModel.cs:
```diff
@@ -1435,19 +1435,22 @@ private CSharpSyntaxNode GetBindingRoot(CSharpSyntaxNode node)
-                }
+                // PROTOTYPE(nullable-api): This doesn't work for speculative models: the root is just the statement
+                // or expression being speculated on.
+                return Root;
```
>return Root; [](start = 16, length = 12)

Can this affect error recovery binding for lambdas? For example, ```GetBoundLambdaOrQuery``` might not expect that we will jump out of an enclosing lambda. Also, it looks like this will "undo" laziness of IOperation tree creation, see ```GetBindingRootOrInitializer````. It feels like a better approach would be to not change the node returned by this function, but instead explicitly bind the root and analyze it when it is appropriate.

333fred replied:
I don't want to lose the debug assert above in nullable-enable code, so I'll add a default parameter to `GetBindingRoot` that will control this condition and only the appropriate places will set it to true.

---
In reply to: [269798837](https://github.com/dotnet/roslyn/pull/34429#discussion_r269798837) [](ancestors = 269798837)

AlekseyTs commented on src/Compilers/CSharp/Portable/Compilation/MemberSemanticModel.cs:
```diff
@@ -1435,19 +1435,22 @@ private CSharpSyntaxNode GetBindingRoot(CSharpSyntaxNode node)
-                switch (current.Kind())
+            for (CSharpSyntaxNode current = node; current != this.Root; current = current.ParentOrStructuredTriviaParent)
+            {
+                switch (current)
```
>switch (current) [](start = 16, length = 16)

It is not clear what has motivated the refactoring inside this loop. It doesn't look like there is an intent to change a behavior. Consider reverting.

333fred replied:
There was initially an intent to change behavior, and this refactoring made that much clearer. After that was ruled out, I found that this version was clearer than the existing code, so I kept it.

---
In reply to: [269801499](https://github.com/dotnet/roslyn/pull/34429#discussion_r269801499) [](ancestors = 269801499)

AlekseyTs commented on src/Compilers/CSharp/Portable/Compilation/CSharpSemanticModel.cs:
```diff
@@ -941,9 +947,8 @@ internal CSharpTypeInfo GetSpeculativeTypeInfoWorker(int position, ExpressionSyn
-            Binder binder;
             ImmutableArray<Symbol> crefSymbols;
-            BoundNode boundNode = GetSpeculativelyBoundExpression(position, expression, bindingOption, out binder, out crefSymbols); //calls CheckAndAdjustPosition
+            BoundNode boundNode = GetSpeculativelyBoundExpression(position, expression, bindingOption, out _, out crefSymbols); //calls CheckAndAdjustPosition
```
>BoundNode boundNode = GetSpeculativelyBoundExpression(position, expression, bindingOption, out _, out crefSymbols); //calls CheckAndAdjustPosition [](start = 12, length = 146)

It doesn't look like any nullable analysis is going to be done on this node. We should also check other ```GetSpeculative...``` APIs.

333fred replied:
Speculative semantic models aren't really handled in this pr, other than basic "support" for paths that go through `GetBoundNodes`. That will require more work in a different PR.

---
In reply to: [269810929](https://github.com/dotnet/roslyn/pull/34429#discussion_r269810929) [](ancestors = 269810929)

AlekseyTs replied:
> Speculative semantic models aren't really handled in this pr, other than basic "support" for paths that go through GetBoundNodes. That will require more work in a different PR.

Would it make sense to to not do any analysis for speculative scenarios for now, even in context of member semantic model?

---
In reply to: [271518790](https://github.com/dotnet/roslyn/pull/34429#discussion_r271518790) [](ancestors = 271518790,269810929)

333fred replied:
I'll add a check in `EnsureBoundRootIfNecessary` to bail out if we're in a speculative model for now.

---
In reply to: [271520509](https://github.com/dotnet/roslyn/pull/34429#discussion_r271520509) [](ancestors = 271520509,271518790,269810929)

AlekseyTs commented on src/Compilers/CSharp/Portable/Compilation/MemberSemanticModel.cs:
```diff
@@ -1833,10 +1836,40 @@ internal ImmutableArray<BoundNode> GetBoundNodes(CSharpSyntaxNode node)
+                    // PROTOTYPE(nullable-api): Handle other symbol types
+                    using (_nodeMapLock.DisposableWrite())
+                    {
+                        _ = GuardedAddBoundTreeAndGetBoundNodeFromMap(node, root);
```
>GuardedAddBoundTreeAndGetBoundNodeFromMap(node, root); [](start = 28, length = 54)

I am not sure if it is safe to interrupt the lock between binding and adding the result into the map. What if another thread gets in between?

AlekseyTs replied:
>_ = GuardedAddBoundTreeAndGetBoundNodeFromMap(node, root); [](start = 24, length = 58)

Not capturing the ```results``` on this code path? Is there a specific reason why?

AlekseyTs commented on src/Compilers/CSharp/Portable/Compilation/MemberSemanticModel.cs:
```diff
@@ -1833,10 +1836,41 @@ internal ImmutableArray<BoundNode> GetBoundNodes(CSharpSyntaxNode node)
+            using (_nodeMapLock.DisposableWrite())
+            {
+
+                results = GuardedAddBoundTreeAndGetBoundNodeFromMap(node, root);
```
>results = GuardedAddBoundTreeAndGetBoundNodeFromMap(node, root); [](start = 16, length = 64)

I am not sure if it is safe to interrupt the lock between binding and adding the result into the map. What if another thread gets in between?

333fred replied:
Then we would accidentally do the analysis twice, but I don't think we should add to the map twice. `GuardedAddBoundTreeAndGetBoundNodeFromMap` will only add new nodes to the map if they're not already in there, so the thread that gets interrupted would just do some extra work and then throw the results away.

---
In reply to: [269814702](https://github.com/dotnet/roslyn/pull/34429#discussion_r269814702) [](ancestors = 269814702)

333fred replied:
I'll change this to avoid change existing semantics, but it does feel like this wouldn't be unsafe, just result in more work, which can already happen if two threads happen to get past the initial check for existing results at the same time.

---
In reply to: [271072016](https://github.com/dotnet/roslyn/pull/34429#discussion_r271072016) [](ancestors = 271072016,269814702)

AlekseyTs commented:
>                 GuardedAddBoundTreeForStandaloneSyntax(syntax, bound);

It looks like we are skipping the rewrite here, Analyzers won't get proper nullability information

---
Refers to: src/Compilers/CSharp/Portable/Compilation/MemberSemanticModel.cs:1389 in a989ba3. [](commit_id = a989ba3aabb985b8a53db0caff4d80f47cea85e0, deletion_comment = False)

AlekseyTs commented:
>                 nodes = GuardedAddBoundTreeAndGetBoundNodeFromMap(lambdaOrQuery, boundOuterExpression);

Do we need to do the rewrite here?

---
Refers to: src/Compilers/CSharp/Portable/Compilation/MemberSemanticModel.cs:1593 in a989ba3. [](commit_id = a989ba3aabb985b8a53db0caff4d80f47cea85e0, deletion_comment = False)

AlekseyTs commented:
>                 nodes = GuardedAddBoundTreeAndGetBoundNodeFromMap(lambdaOrQuery, boundOuterExpression);

Do we need to do the rewrite here?

---
Refers to: src/Compilers/CSharp/Portable/Compilation/MemberSemanticModel.cs:1621 in a989ba3. [](commit_id = a989ba3aabb985b8a53db0caff4d80f47cea85e0, deletion_comment = False)

cston commented on src/Compilers/CSharp/Portable/Compilation/CSharpSemanticModel.cs:
```diff
@@ -2058,13 +2069,16 @@ private static bool IsUserDefinedTrueOrFalse(BoundUnaryOperator @operator)
 
             return CSharpTypeInfo.None;
+
+            (TypeSymbol, NullabilityInfo) getTypeAndNullability(BoundExpression expr) => (expr.Type, expr.TopLevelNullability);
```
>( [](start = 12, length = 1)

`static`

cston commented on src/Compilers/CSharp/Portable/Compilation/MemberSemanticModel.cs:
```diff
@@ -1833,10 +1836,41 @@ internal ImmutableArray<BoundNode> GetBoundNodes(CSharpSyntaxNode node)
+                // Always rewrite in DEBUG mode and just throw the results away.
+                if (MemberSymbol is MethodSymbol method)
+                {
+                    _ = NullableWalker.AnalyzeAndRewrite(Compilation, method, Bind(incrementalBinder, Root, _ignoredDiagnostics), RootBinder.Conversions, _ignoredDiagnostics);
```
>_ignoredDiagnostics [](start = 108, length = 19)

Perhaps create a `new DiagnosticBag()` for this codepath, to minimize the observable differences in `DEBUG`. (Even though the current `DiagnosticBag` is named `_ignoredDiagnostics`, it is stored in a field so some member might look at the value later.)

cston commented on src/Compilers/CSharp/Portable/SymbolDisplay/SymbolDisplay.cs:
```diff
@@ -127,26 +156,50 @@ public static class SymbolDisplay
         {
-            format = format ?? SymbolDisplayFormat.MinimallyQualifiedFormat;
-            return ToDisplayParts(symbol, topLevelNullability, semanticModel, position, format, minimal: true);
+            // PROTOTYPE(nullable-api): Refactor this. We need to be able to handle non-TypeSymbol inputs, and add tests
```
> and add tests [](start = 106, length = 14)

Please consider moving this comment to the API, if there's not already a PROTOTYPE comment for testing the API.

AlekseyTs commented on src/Compilers/CSharp/Portable/BoundTree/NullabilityRewriter.cs:
```diff
@@ -8,5 +13,68 @@ protected override BoundExpression VisitExpressionWithoutStackGuard(BoundExpress
+
+            while (true)
+            {
+                if (child is BoundBinaryOperatorBase childBinary)
```
>child is BoundBinaryOperatorBase childBinary [](start = 20, length = 44)

It looks like this condition can simply be moved into the ```while``` condition and the ```if``` can be removed. Also, it looks like we are always doing at least one iteration, so the first condition check is redundant and it can be avoided by using a loop with post condition. Consider the following refactoring:
```
BoundBinaryOperatorBase currentBinary = binaryOperator;
// Remove            BoundExpression child = binaryOperator;

do
{
    stack.Push(currentBinary);
    currentBinary= currentBinary.Left as BoundBinaryOperatorBase;
}
while (currentBinary != null);
```

AlekseyTs commented on src/Compilers/CSharp/Portable/BoundTree/NullabilityRewriter.cs:
```diff
@@ -8,5 +13,68 @@ protected override BoundExpression VisitExpressionWithoutStackGuard(BoundExpress
+            }
+
+            Debug.Assert(stack.Count > 0);
+            currentBinary = null;
```
>currentBinary = null; [](start = 12, length = 21)

This becomes redundant after the refactoring suggested above and can be removed.

AlekseyTs commented on src/Compilers/CSharp/Portable/BoundTree/NullabilityRewriter.cs:
```diff
@@ -8,5 +13,68 @@ protected override BoundExpression VisitExpressionWithoutStackGuard(BoundExpress
+
+            while (stack.Count > 0)
+            {
+                var left = currentBinary;
```
>var left = currentBinary; [](start = 16, length = 25)

This feels more complicated than it needs to be. On the first iteration ```currentBinary``` is null. So, the ```updateNode``` has to have conditional logic to special case this and that logic will be executed on every iteration. It is not clear what is the benefit having that local function at all. It is used in one place and its body can simply be inlined. Also, we know that at least one iteration will  be done, checking condition for the first iteration is redundant and can be avoided by switching to a loop with post condition. To summarize all this, consider the following refactoring:
```
BoundExpression child = (BoundExpression)rewriter.Visit(stack.Peek().Left);

do
{
    currentBinary = stack.Pop();

                var foundInfo = rewriter._updatedNullabilities.TryGetValue(currentBinary, out (NullabilityInfo Info, TypeSymbol Type) infoAndType);
                var right = (BoundExpression)rewriter.Visit(currentBinary.Right);
                currentBinary = nodeUpdater(currentBinary, child, right, foundInfo ? infoAndType.Type : currentBinary.Type);

                if (foundInfo)
                {
                    currentBinary.TopLevelNullability = infoAndType.Info;
                }

     child = currentBinary;
}
while (stack.Count > 0);
```

AlekseyTs commented on src/Compilers/CSharp/Portable/BoundTree/NullabilityRewriter.cs:
```diff
@@ -8,5 +13,68 @@ protected override BoundExpression VisitExpressionWithoutStackGuard(BoundExpress
+
+        public override BoundNode VisitBinaryOperator(BoundBinaryOperator node)
+        {
+            return VisitBinaryOperatorBase(node, (binary, left, right, type) => node.Update(node.OperatorKind, node.ConstantValueOpt, node.MethodOpt, node.ResultKind, left, right, type));
```
>(binary, left, right, type) => node.Update(node.OperatorKind, node.ConstantValueOpt, node.MethodOpt, node.ResultKind, left, right, type) [](start = 49, length = 136)

This feels wrong. Instead of updating the ```binary``` the lambda will keep updating the original root node over and over again. So, none of the BoundBinaryOperatorBase children collected into the stack below are going to be updated properly. Also, given that we are supposed to be able to update ```BoundBinaryOperator``` and ```BoundUserDefinedConditionalLogicalOperator``` nodes, we cannot rely on the same ```Update``` method to do that. Since none of the issues were caught by the testing, it feels like we are lacking adequate test coverage for the new code in this file.

AlekseyTs commented on src/Compilers/CSharp/Portable/BoundTree/NullabilityRewriter.cs:
```diff
@@ -8,5 +13,68 @@ protected override BoundExpression VisitExpressionWithoutStackGuard(BoundExpress
+
+        public override BoundNode VisitUserDefinedConditionalLogicalOperator(BoundUserDefinedConditionalLogicalOperator node)
+        {
+            return VisitBinaryOperatorBase(node, (binary, left, right, type) => node.Update(node.OperatorKind, node.LogicalOperator, node.TrueOperator, node.FalseOperator, node.ResultKind, left, right, type));
```
>(binary, left, right, type) => node.Update(node.OperatorKind, node.LogicalOperator, node.TrueOperator, node.FalseOperator, node.ResultKind, left, right, type) [](start = 49, length = 158)

It looks like this lambda has all the same problems as the one in the function above.

AlekseyTs commented on src/Compilers/CSharp/Portable/Symbols/TypeParameterSymbol.cs:
```diff
@@ -690,10 +690,13 @@ internal override TypeSymbol MergeNullability(TypeSymbol other, VarianceKind var
         {
-            get => default;
-        }
+            false when !HasReferenceTypeConstraint => CodeAnalysis.NullableAnnotation.NotApplicable,
```
>!HasReferenceTypeConstraint [](start = 23, length = 27)

One ```false``` case with conditional inside might be easier to follow.

333fred replied:
I felt that this was clear and easy to follow.

---
In reply to: [270178845](https://github.com/dotnet/roslyn/pull/34429#discussion_r270178845) [](ancestors = 270178845)

AlekseyTs commented on src/Compilers/CSharp/Portable/Symbols/TypeWithAnnotations.cs:
```diff
@@ -304,15 +304,14 @@ public string ToDisplayString(SymbolDisplayFormat format = null)
             {
                 if (format.MiscellaneousOptions.IncludesOption(SymbolDisplayMiscellaneousOptions.IncludeNullableReferenceTypeModifier) &&
-                    !IsNullableType() && !Type.IsValueType &&
+                    (!HasType || !IsNullableType() && !Type.IsValueType) &&
```
>!IsNullableType() && !Type.IsValueType [](start = 33, length = 38)

Consider adding parens for readability

AlekseyTs commented on src/Compilers/CSharp/Portable/Symbols/TypeWithAnnotations.cs:
```diff
@@ -304,15 +304,14 @@ public string ToDisplayString(SymbolDisplayFormat format = null)
-                    !Type.IsValueType &&
-                    NullableAnnotation.IsNotAnnotated() &&
-                    !Type.IsTypeParameterDisallowingAnnotation())
+                    (!HasType || !Type.IsValueType && !Type.IsTypeParameterDisallowingAnnotation()) &&
```
>!Type.IsValueType && !Type.IsTypeParameterDisallowingAnnotation() [](start = 33, length = 65)

Consider adding parens for readability

AlekseyTs commented on src/Compilers/CSharp/Portable/Symbols/TypeParameterSymbol.cs:
```diff
@@ -690,10 +690,13 @@ internal override TypeSymbol MergeNullability(TypeSymbol other, VarianceKind var
 
-        CodeAnalysis.NullableAnnotation ITypeParameterSymbol.ReferenceTypeConstraintNullableAnnotation
+        CodeAnalysis.NullableAnnotation ITypeParameterSymbol.ReferenceTypeConstraintNullableAnnotation => ReferenceTypeConstraintIsNullable switch
         {
```
>{ [](start = 8, length = 1)

Consider indenting the ```switch``` body and putting it on a separate line. Like so:
```
        CodeAnalysis.NullableAnnotation ITypeParameterSymbol.ReferenceTypeConstraintNullableAnnotation =>
            ReferenceTypeConstraintIsNullable switch
            {
                false when !HasReferenceTypeConstraint => CodeAnalysis.NullableAnnotation.NotApplicable,
                false => CodeAnalysis.NullableAnnotation.NotAnnotated,
                true => CodeAnalysis.NullableAnnotation.Annotated,
                null => CodeAnalysis.NullableAnnotation.Disabled,
            };
```

I think this way the code will be easier to follow.

333fred replied:
I didn't do this because currently the IDE doesn't have a good formatting story for switch expressions, and that won't be fixed until 16.1 brings in @sharwell's change in [https://github.com/dotnet/roslyn/pull/33938](https://github.com/dotnet/roslyn/pull/33938). I can format them like this and disable the formatting analyzer for this region if you'd prefer.

---
In reply to: [270182494](https://github.com/dotnet/roslyn/pull/34429#discussion_r270182494) [](ancestors = 270182494)

AlekseyTs commented on src/Compilers/CSharp/Portable/Symbols/NullableFlowStateExtensions.cs:
```diff
@@ -23,23 +22,20 @@ internal static class NullableFlowStateExtensions
 
-        internal static CodeAnalysis.NullableFlowState ToPublicFlowState(this NullableFlowState nullableFlowState)
+        internal static CodeAnalysis.NullableFlowState ToPublicFlowState(this NullableFlowState nullableFlowState) => nullableFlowState switch
         {
```
>{ [](start = 8, length = 1)

Consider indenting the ```switch``` body and putting it (the ```switch```) on a separate line.

AlekseyTs commented on src/Compilers/CSharp/Portable/Symbols/NullableFlowStateExtensions.cs:
```diff
@@ -23,23 +22,20 @@ internal static class NullableFlowStateExtensions
         // PROTOTYPE(nullable-api): remove if possible
-        public static NullableFlowState ToInternalFlowState(this CodeAnalysis.NullableFlowState flowState)
+        public static NullableFlowState ToInternalFlowState(this CodeAnalysis.NullableFlowState flowState) => flowState switch
         {
```
>{ [](start = 8, length = 1)

Here as well.

AlekseyTs commented on src/Compilers/CSharp/Portable/Symbols/NullableAnnotationExtensions.cs:
```diff
@@ -47,10 +48,38 @@ public static NullableAnnotation EnsureCompatible(this NullableAnnotation a, Nul
 
-        internal static CodeAnalysis.NullableAnnotation ToPublicAnnotation(this NullableAnnotation annotation)
+        internal static CodeAnalysis.NullableAnnotation ToPublicAnnotation(this NullableAnnotation annotation) => annotation switch
         {
```
>{ [](start = 8, length = 1)

Consider indenting the ```switch``` body and putting the ```switch``` on a separate line.

AlekseyTs commented on src/Compilers/CSharp/Portable/Symbols/NullableAnnotationExtensions.cs:
```diff
@@ -47,10 +48,38 @@ public static NullableAnnotation EnsureCompatible(this NullableAnnotation a, Nul
         }
+
+        internal static NullableAnnotation ToInternalAnnotation(this CodeAnalysis.NullableAnnotation annotation) => annotation switch
+        {
```
>{ [](start = 8, length = 1)

Consider indenting the ```switch``` body and putting the ```switch``` on a separate line.

AlekseyTs commented on src/Compilers/CSharp/Portable/Symbols/NullableAnnotationExtensions.cs:
```diff
@@ -47,10 +48,38 @@ public static NullableAnnotation EnsureCompatible(this NullableAnnotation a, Nul
+            return internalAnnotation.ToNullabilityInfo(type);
+        }
+
+        internal static NullabilityInfo ToNullabilityInfo(this NullableAnnotation annotation, TypeSymbol type)
```
>NullableAnnotation [](start = 63, length = 18)

How is this ```NullableAnnotation``` different from ```CodeAnalysis.NullableAnnotation```? Are these different types? Consider explicitly specifying namespace in this case too, the code will be easier to follow.

AlekseyTs replied:
In general, I think we should avoid having types with same names in different namespaces, with exception of VB vs. C#.

---
In reply to: [270184841](https://github.com/dotnet/roslyn/pull/34429#discussion_r270184841) [](ancestors = 270184841)

cston replied:
These are public and internal versions of the same state. And we should have good names for both. For most consumers (other than the translation in SemanticModel), there should not be any ambiguity.

---
In reply to: [270186019](https://github.com/dotnet/roslyn/pull/34429#discussion_r270186019) [](ancestors = 270186019,270184841)

333fred replied:
The public version has a default state that allows the disambiguation of C# 7 expressions and things that don't have the concept of nullability, such as statements. I had originally renamed the internal version of a type like we have `TypeInfo` and `CSharpTypeInfo`, but I received pushback on that and did this instead.

---
In reply to: [270189832](https://github.com/dotnet/roslyn/pull/34429#discussion_r270189832) [](ancestors = 270189832,270186019,270184841)

AlekseyTs commented on src/Compilers/Test/Utilities/CSharp/CompilationTestUtils.cs:
```diff
@@ -306,31 +306,20 @@ internal static void VerifyTypes(this CSharpCompilation compilation, SyntaxTree
                 var expectedTypes = annotations.SelectAsArray(annotation => annotation.Text);
-                var actualTypes = annotations.SelectAsArray(annotation => toDisplayString(annotation.Expression));
+                var actualTypes = annotations.SelectAsArray(annotation =>
+                {
```
>{ [](start = 16, length = 1)

Please indent the body of the lambda

AlekseyTs commented:
Done with review pass (iteration 3), new tests are not looked at.

333fred commented:
>                 GuardedAddBoundTreeForStandaloneSyntax(syntax, bound);

I don't think this is the right place to be doing any kind of rewriting. It seems like the consumer of this API should be the one doing the rewrite. There's only place that will actually hit this, `MethodCompiler.CompileMethod`. I think fully tracking and testing a change there is outside the scope of this PR, so I'll leave a prototype comment and save it for a followup.

---
In reply to: [477684819](https://github.com/dotnet/roslyn/pull/34429#issuecomment-477684819) [](ancestors = 477684819)

---
Refers to: src/Compilers/CSharp/Portable/Compilation/MemberSemanticModel.cs:1389 in a989ba3. [](commit_id = a989ba3aabb985b8a53db0caff4d80f47cea85e0, deletion_comment = False)

333fred commented:
>                 nodes = GuardedAddBoundTreeAndGetBoundNodeFromMap(lambdaOrQuery, boundOuterExpression);

I think, if I'm reading the code correctly, I can just call my forthcoming `EnsureRootBoundForNullabilityIfNecessary` before we first try to get bound nodes, and that should bind the entire method, including any bound or unbound lambdas. We do need to the analysis in the error cases below, but I don't think we'll get here for non error cases.

---
In reply to: [477685514](https://github.com/dotnet/roslyn/pull/34429#issuecomment-477685514) [](ancestors = 477685514)

---
Refers to: src/Compilers/CSharp/Portable/Compilation/MemberSemanticModel.cs:1593 in a989ba3. [](commit_id = a989ba3aabb985b8a53db0caff4d80f47cea85e0, deletion_comment = False)

AlekseyTs commented:
>                 nodes = GuardedAddBoundTreeAndGetBoundNodeFromMap(lambdaOrQuery, boundOuterExpression);

> We do need to the analysis in the error cases below, but I don't think we'll get here for non error cases.

Are you saying we don't need to do the analysis here even though we bind some syntax? 

---
In reply to: [479230801](https://github.com/dotnet/roslyn/pull/34429#issuecomment-479230801) [](ancestors = 479230801,477685514)

---
Refers to: src/Compilers/CSharp/Portable/Compilation/MemberSemanticModel.cs:1593 in a989ba3. [](commit_id = a989ba3aabb985b8a53db0caff4d80f47cea85e0, deletion_comment = False)

AlekseyTs commented on src/Compilers/CSharp/Portable/BoundTree/NullabilityRewriter.cs:
```diff
@@ -8,5 +14,58 @@ protected override BoundExpression VisitExpressionWithoutStackGuard(BoundExpress
+            {
+                stack.Push(currentBinary);
+                currentBinary = currentBinary.Left as BoundBinaryOperatorBase;
+            } while (currentBinary != null);
```
>} while (currentBinary != null); [](start = 12, length = 32)

Is placing ```while``` on the same line as the closing brace (```}```) follows our coding style?

333fred replied:
Well, the formatting analyzer doesn't complain about it, but by allman bracing styles it probably should.

---
In reply to: [271539092](https://github.com/dotnet/roslyn/pull/34429#discussion_r271539092) [](ancestors = 271539092)

AlekseyTs commented on src/Compilers/CSharp/Portable/BoundTree/NullabilityRewriter.cs:
```diff
@@ -8,5 +14,58 @@ protected override BoundExpression VisitExpressionWithoutStackGuard(BoundExpress
+                var type = foundInfo ? infoAndType.Type : currentBinary.Type;
+
+                currentBinary = currentBinary switch
+                {
```
>{ [](start = 16, length = 1)

Please indent the ```switch``` body.

AlekseyTs commented on src/Compilers/CSharp/Portable/BoundTree/NullabilityRewriter.cs:
```diff
@@ -8,5 +14,58 @@ protected override BoundExpression VisitExpressionWithoutStackGuard(BoundExpress
+
+                currentBinary = currentBinary switch
+                {
+                    BoundBinaryOperator binary => (BoundBinaryOperatorBase)binary.Update(binary.OperatorKind, binary.ConstantValueOpt, binary.MethodOpt, binary.ResultKind, leftChild, right, type),
```
>binary.MethodOpt [](start = 135, length = 16)

It feels like we might want to adjust the symbol as well.

333fred replied:
We definitely will, but we're not currently doing that anywhere. We'll need to do this for pretty much any bound node that has an internal symbol, like BoundCalls, BoundPropertyAccesses, etc.

---
In reply to: [271540045](https://github.com/dotnet/roslyn/pull/34429#discussion_r271540045) [](ancestors = 271540045)

AlekseyTs replied:
> We definitely will, but we're not currently doing that anywhere. We'll need to do this for pretty much any bound node that has an internal symbol, like BoundCalls, BoundPropertyAccesses, etc.

Consider adding a prototype comment

---
In reply to: [273229265](https://github.com/dotnet/roslyn/pull/34429#discussion_r273229265) [](ancestors = 273229265,271540045)

AlekseyTs commented on src/Compilers/CSharp/Portable/BoundTree/NullabilityRewriter.cs:
```diff
@@ -8,5 +14,58 @@ protected override BoundExpression VisitExpressionWithoutStackGuard(BoundExpress
+                currentBinary = currentBinary switch
+                {
+                    BoundBinaryOperator binary => (BoundBinaryOperatorBase)binary.Update(binary.OperatorKind, binary.ConstantValueOpt, binary.MethodOpt, binary.ResultKind, leftChild, right, type),
+                    BoundUserDefinedConditionalLogicalOperator logical => logical.Update(logical.OperatorKind, logical.LogicalOperator, logical.TrueOperator, logical.FalseOperator, logical.ResultKind, leftChild, right, type),
```
>logical.LogicalOperator, logical.TrueOperator, logical.FalseOperator [](start = 111, length = 68)

It feels like we might want to adjust the symbols as well.

AlekseyTs commented on src/Compilers/CSharp/Portable/BoundTree/NullabilityRewriter.cs:
```diff
@@ -8,5 +14,58 @@ protected override BoundExpression VisitExpressionWithoutStackGuard(BoundExpress
+                }
+
+                leftChild = currentBinary;
+            } while (stack.Count > 0);
```
>} while (stack.Count > 0); [](start = 12, length = 26)

Same style question.

cston commented on src/Compilers/CSharp/Portable/Compilation/MemberSemanticModel.cs:
```diff
@@ -1615,6 +1622,20 @@ private BoundNode GetBoundLambdaOrQuery(CSharpSyntaxNode lambdaOrQuery)
+                {
+                    var diagnostics = new DiagnosticBag();
+                    _ = RewriteNullableBoundNodes(boundOuterExpression, incrementalBinder.Conversions, diagnostics);
+                    diagnostics.Free();
```
>diagnostics.Free(); [](start = 20, length = 19)

This instance wasn't allocated from the pool.

cston commented on src/Compilers/CSharp/Portable/Compilation/MemberSemanticModel.cs:
```diff
@@ -1788,6 +1809,73 @@ private static Binder GetLambdaEnclosingBinder(int position, CSharpSyntaxNode st
+                )
+            {
+#if DEBUG
+                if (!EnableNullableAnalysis) diagnostics.Free();
```
>diagnostics.Free(); [](start = 45, length = 19)

This instance wasn't allocated from the pool so this `#if DEBUG` region can be removed.

cston commented on src/Compilers/CSharp/Portable/Compilation/MemberSemanticModel.cs:
```diff
@@ -1788,6 +1809,73 @@ private static Binder GetLambdaEnclosingBinder(int position, CSharpSyntaxNode st
+            {
+                diagnostics.Free();
+            }
+#endif
```
Can be removed.

cston commented on src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs:
```diff
@@ -1560,9 +1556,17 @@ private TypeWithState VisitRefExpression(BoundExpression expr, TypeWithAnnotatio
+            var nullableReturnType = (_methodSignatureOpt ?? method)?.ReturnTypeWithAnnotations;
+            Debug.Assert(!nullableReturnType.HasValue || (object)nullableReturnType.Value != LambdaSymbol.ReturnTypeIsBeingInferred);
+
+            if (!nullableReturnType.HasValue)
```
>nullableReturnType [](start = 17, length = 18)

The `TypeWithAnnotations?` variable seems unnecessary. Consider simply returning early if the method is `null`.
```C#
var method = _methodSignatureOpt ?? (_symbol as MethodSymbol);
if (method is null)
{
    type = default;
    return false;
}

var returnType = method.ReturnTypeWithAnnotations;
```

AlekseyTs commented on src/Compilers/CSharp/Portable/Compilation/MemberSemanticModel.cs:
```diff
@@ -1588,6 +1591,10 @@ private BoundNode GetBoundLambdaOrQuery(CSharpSyntaxNode lambdaOrQuery)
                 BoundNode boundOuterExpression = this.Bind(incrementalBinder, nodeToBind, _ignoredDiagnostics);
                 nodes = GuardedAddBoundTreeAndGetBoundNodeFromMap(lambdaOrQuery, boundOuterExpression);
+
+                // If we're here, nullable analysis must be off, or we're in an error scenario and didn't find any
```
>// If we're here, nullable analysis must be off, or we're in an error scenario and didn't find any [](start = 16, length = 98)

I think we should have a prototype comment here instead of an assert. We need to make a decision what to do about code that we bind in error recovery situations and then adjust this place accordingly. If we can ever reach this code with ```!nodes.IsDefaultOrEmpty```, the I have no doubt it can be reached with ```EnableNullableAnalysis``` at the same time. So the assert below is simply indicating a possible test gap rather than the fact that it is safe to do nothing here.

AlekseyTs commented on src/Compilers/CSharp/Portable/Compilation/MemberSemanticModel.cs:
```diff
@@ -1615,6 +1622,20 @@ private BoundNode GetBoundLambdaOrQuery(CSharpSyntaxNode lambdaOrQuery)
+
+                if (EnableNullableAnalysis)
+                {
+                    boundOuterExpression = RewriteNullableBoundNodes(boundOuterExpression, incrementalBinder.Conversions, _ignoredDiagnostics);
```
>boundOuterExpression = RewriteNullableBoundNodes(boundOuterExpression, incrementalBinder.Conversions, _ignoredDiagnostics); [](start = 20, length = 123)

Do we have test that makes this call observable?

AlekseyTs commented on src/Compilers/CSharp/Portable/Compilation/MemberSemanticModel.cs:
```diff
@@ -1788,6 +1809,73 @@ private static Binder GetLambdaEnclosingBinder(int position, CSharpSyntaxNode st
+
+            upgradeableLock.EnterWrite();
+
+            Debug.Assert(Root == GetBindableSyntaxNode(Root));
```
>GetBindableSyntaxNode(Root) [](start = 33, length = 27)

For some reason the function below explicitly calls ```GetBindableSyntaxNode(Root)``` for production code, I think it would be safer if we do the same in this function and simply operate on the result we get back rather than using ```Root``` explicitly and assume that the nodes are the same.

AlekseyTs commented:
Done with review pass (iteration 4)

333fred commented:
@AlekseyTs @cston addressed feedback. A separate PR is working on bringing in the infra changes necessary to get this branch to build, and I'll merge that in here when it's done.

---
In reply to: [480947732](https://github.com/dotnet/roslyn/pull/34429#issuecomment-480947732) [](ancestors = 480947732)

AlekseyTs commented on src/Compilers/CSharp/Portable/Compilation/MemberSemanticModel.cs:
```diff
@@ -1788,6 +1802,64 @@ private static Binder GetLambdaEnclosingBinder(int position, CSharpSyntaxNode st
+            // Currently, it will return incorrect info because it's just running analysis on the speculated
+            // part.
+            var binder = GetEnclosingBinder(GetAdjustedNodePosition(GetBindableSyntaxNode(Root)));
+            var boundRoot = Bind(binder, Root, diagnostics);
```
>Root [](start = 41, length = 4)

```GetBindableSyntaxNode(Root)```?

AlekseyTs commented on src/Compilers/CSharp/Portable/Compilation/MemberSemanticModel.cs:
```diff
@@ -1788,6 +1802,64 @@ private static Binder GetLambdaEnclosingBinder(int position, CSharpSyntaxNode st
+
+            using var upgradeableLock = _nodeMapLock.DisposableUpgradeableRead();
+
+            if (_guardedNodeMap.ContainsKey(Root)
```
>Root [](start = 44, length = 4)

```GetBindableSyntaxNode(Root)```?

AlekseyTs commented on src/Compilers/CSharp/Portable/Compilation/MemberSemanticModel.cs:
```diff
@@ -1788,6 +1802,64 @@ private static Binder GetLambdaEnclosingBinder(int position, CSharpSyntaxNode st
+            var boundRoot = Bind(binder, Root, diagnostics);
+            boundRoot = RewriteNullableBoundNodes(boundRoot, binder.Conversions, diagnostics);
+
+            if (EnableNullableAnalysis)
```
>EnableNullableAnalysis [](start = 16, length = 22)

``` && !IsSpeculativeSemanticModel``` ?

AlekseyTs commented:
Done with review pass (iteration 5)

333fred commented:
@AlekseyTs resolved feedback. Commit 6 has the feedback, 7 is a merge from upstream to get CI building again.

---
In reply to: [481440265](https://github.com/dotnet/roslyn/pull/34429#issuecomment-481440265) [](ancestors = 481440265)

AlekseyTs commented on src/Compilers/CSharp/Portable/Compilation/MemberSemanticModel.cs:
```diff
@@ -1788,6 +1802,65 @@ private static Binder GetLambdaEnclosingBinder(int position, CSharpSyntaxNode st
+
+            if (EnableNullableAnalysis && !IsSpeculativeSemanticModel)
+            {
+                GuardedAddBoundTreeForStandaloneSyntax(Root, boundRoot);
```
>Root [](start = 55, length = 4)

```bindableRoot```?

AlekseyTs commented:
Done with review pass (iteration 8)

cston commented on src/Dependencies/PooledObjects/ArrayBuilder.cs:
```diff
@@ -9,7 +9,7 @@ namespace Microsoft.CodeAnalysis.PooledObjects
     [DebuggerDisplay("Count = {Count,nq}")]
     [DebuggerTypeProxy(typeof(ArrayBuilder<>.DebuggerProxy))]
-    internal sealed partial class ArrayBuilder<T> : IReadOnlyCollection<T>, IReadOnlyList<T>
+    internal sealed partial class ArrayBuilder<T> : IReadOnlyCollection<T>, IReadOnlyList<T>, IDisposable
```
>IDisposable [](start = 94, length = 11)

Are we using this interface anywhere? I'm not against the change but allowing `using` statements for builders has been suggested (unsuccessfully) before.

333fred replied:
Whoops, I was experimenting with this, I thought in a different repo. Must have missed it.

---
In reply to: [274100294](https://github.com/dotnet/roslyn/pull/34429#discussion_r274100294) [](ancestors = 274100294)
